builtin/logical/database/path_roles.go

```diff
@@ -482,19 +504,32 @@ func (b *databaseBackend) pathStaticRoleCreateUpdate(ctx context.Context, req *l
 
 	// Only call setStaticAccount if we're creating the role for the
 	// first time
+	var item *queue.Item
 	switch req.Operation {
 	case logical.CreateOperation:
 		// setStaticAccount calls Storage.Put and saves the role to storage
 		resp, err := b.setStaticAccount(ctx, req.Storage, &setStaticAccountInput{
-			RoleName:   name,
-			Role:       role,
-			CreateUser: createRole,
+			RoleName: name,
+			Role:     role,
 		})
 		if err != nil {
-			return nil, err
+			if resp != nil && resp.WALID != "" {
+				b.Logger().Debug("deleting WAL for failed role creation", "WAL ID", resp.WALID, "role", name)
+				walDeleteErr := framework.DeleteWAL(ctx, req.Storage, resp.WALID)
+				if walDeleteErr != nil {
+					b.Logger().Debug("failed to delete WAL for failed role creation", "WAL ID", resp.WALID, "error", walDeleteErr)
+					var merr *multierror.Error
+					merr = multierror.Append(merr, err)
+					merr = multierror.Append(merr, fmt.Errorf("failed to clean up WAL from failed role creation: %w", walDeleteErr))
+					err = merr.ErrorOrNil()
+				}
+			}
```

Comment:
I don't know if this applies to OpenLDAP as well, but we're swallowing an error here. For example, I had configured my connection with `allowed_roles=["my-role"]`, was trying to set up a static role `foo-role`, was getting success messages on the Write operation but "role not found" in the server logs and no role actually created. Turned out the Write was actually returning `"'foo-role' is not an allowed role"` but that was being swallowed.

Reply:
Thanks! Fixed: https://github.com/hashicorp/vault/pull/12563/commits/b1c8a06e10c2b35379d16241c0f8f2c542d117cc